generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -569,6 +571,54 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
         fuzzer)));
   }
 
+  /**
+   * Function to create an opaque zero (specifically an opaque zero ved3) by taking the cross
+   * product of two equivalent vec3s. This opaque function relies on the fact that if u == v, where
+   * u and v are vec3, then cross(u, v) = (0, 0, 0). These equivalent vec3s are produced by
+   * fuzzing random float values and applying different identity functions to them per vector.
+   * @param type - the base type of the opaque value being created.
+   * @param constContext - true if we're in a constant expression context, false otherwise.
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque zero vec3
+   *     made from the cross product of two equal vec3s.
+   */
+  private Optional<Expr> opaqueZeroVectorCross(BasicType type, boolean constContext,
+                                               final int depth,
+                                               Fuzzer fuzzer, boolean isZero) {
+    assert isZero;
+    if (type != BasicType.VEC3) {
+      // Cross product only works on vec3 and can only produce vec3 as a return type.
+      return Optional.empty();
+    }
+    // We only want literals, so we need to make our own LiteralFuzzer rather than use the supplied
+    // Fuzzer object.
+    final LiteralFuzzer litFuzzer = new LiteralFuzzer(generator);
+    // cross(vec3(firstVec3ConstructorArgs), vec3(secondVec3ConstructorArgs))
+    final List<Expr> firstVec3ConstructorArgs = new ArrayList<Expr>();
+    final List<Expr> secondVec3ConstructorArgs = new ArrayList<Expr>();
+    for (int i = 0; i < type.getNumElements(); i++) {
+      final Optional<Expr> maybeFuzzedFloatLiteral = litFuzzer.fuzz(type.getElementType());
```

Comment:
I would change to:

// We are guarenteed to obtain a fuzzed expression, since the element type for a vec3 is float, and it is always possible to fuzz a float literal.

Reply:
Fixed. I had trouble trying to phrase this comment, so thanks for adding.